### PR TITLE
Add regex capture and rename capture_bytes to capture

### PR DESCRIPTION
**Pull request by XAMPPRocky**
This PR adds a regex capture strategy to the CaptureBytes filter, and renames it from `CaptureBytes` to capture, to better reflect the new `Capture` trait contract, where a capture strategy can return any valid `metadata::Value` instead of only `Bytes`. This also adds a `/is_present` extension metadata key which is appended ontop of whatever the `metadata_key`  is, this will be useful in combination with #447 to allow to match on whether any successful capture occured, rather than only matching values.

### CODE REVIEW

File: docs/src/filters/token_router.md
Comment (iffyio) [inline docs/src/filters/token_router.md:90]: is it possible to use the [proposed schema here](https://github.com/googleforgames/quilkin/pull/387#issuecomment-910179544) that groups the fields by strategy? this look like we'll have 1 flat object where all fields are together and we'll have to document which belongs to which strategy

File: src/filters/capture.rs
Comment (iffyio) [inline src/filters/capture.rs:73]: the clone+concat looks expensive, we should be able to create the string once on instantiation since its not changing?

Can you elaborate what `/is_present` is for I didn't understand from the PR description and it wasnt clear looking at the code?

File: src/filters/capture/regex.rs
Comment (iffyio) [inline src/filters/capture/regex.rs:25]: this looks like regex can potentially capture a list or single value? if so we should document the behavior (or maybe just return a singleton list in the latter so its consistent?)

File: src/filters/capture.rs
Comment (XAMPPRocky) [reply src/filters/capture.rs:73]: `/is_present` is more useful with the inclusion of the `Match` filter and other conditional flow filters. I wrote an [example](https://github.com/googleforgames/quilkin/discussions/459#discussion-3764877) that uses it, but in essence, it allows you to check for whether a capture succeeded without checking the specific contents.

This is more useful with the inclusion of regex because something like `\d{8}$` would capture the last eight digits, and if you wanted to check if that worked you'd need to match every possible combination of those digits, where as   with `/is_present` we can check it with a couple lines. E.g.

```yaml
  filters:
  - name: quilkin.extensions.filters.capture.v1alpha1.Capture
    config:
      strategy:
        kind: regex
        regex: \d{8}$
  - name: quilkin.extensions.filters.matches.v1alpha1.Matches
    config:
        metadataKey: quilkin.dev/capture_bytes/is_present
        fallthrough: pass
        branches:
        # If present, drop the packet.
        - value: true
          filter: quilkin.extensions.filters.drop.v1alpha1.Drop
```

File: src/filters/capture.rs
Comment (markmandel) [inline src/filters/capture.rs:64]: Just noticed that this key is not documented, so we should add that to `capture.md`.

File: src/filters/capture.rs
Comment (markmandel) [inline src/filters/capture.rs:102]: I like `proto::Capture` better here 👍🏻 Just reads nicer.

File: src/filters/capture.rs
Comment (markmandel) [inline src/filters/capture.rs:285]: I'd like to see a `regex_capture()` test like `end_capture()` and `beginning_capture()`, so we can have a faster interation loop without have to rely on an e2e test to determine if the regex capture is working as expected.

File: tests/capture.rs
Comment (markmandel) [inline tests/capture.rs:38]: I actually didn't realise that you could do neat stuff like this! 😮 (I always relate regex with doing character matching) -- this is very, very cool, I hadn't thought through the implications of this addition.

I think it would be awesome to show in capture.md under "Configuration Examples" that you could do first 3 values capturing using `prefix`, but also do the same thing using `regex` -- two different ways to do the same thing, just to give people ideas of what is possible.

Very nice.

File: docs/src/filters/capture.md
Comment (XAMPPRocky) [reply docs/src/filters/capture.md:24]: > Also wondering if we need to be explicit that a Regex filter that returns more than one value will be dropped by the Token Router.

Well that's why it currently returns a single value, so that it continues to work with token_router. I wonder if we enforce regex to always be a list, that we should also change token_router to check if anything in the list matches?

File: src/filters/capture.rs
Comment (markmandel) [inline src/filters/capture.rs:81]: I just noticed this change in behaviour - this would mean that if there is no match, we drop the whole packet.

I don't think we want to change the behaiviour here. The job of this filter is purely to capture data - it's not to make decisions about it.

🤔 Unless we want to add an option to "drop on failing to match" - which I think would be a nice addition (we could do this in a subsequent PR/track in an issue)?

File: docs/src/filters/capture.md
Comment (markmandel) [reply docs/src/filters/capture.md:24]: Except `Regex` can return a `Value::List` or a single value?

https://github.com/googleforgames/quilkin/blob/2828b8678d6ab3440ddc61eac27e6264f5528bed/src/filters/capture/regex.rs#L20-L26

Or am I reading `matches.into_iter().next()` wrong?

> that we should also change token_router to check if anything in the list matches?

There's probably an interesting discussion of having an option on `token_router` to allow matching against lists of values - shall we create an issue for it? (seems to me like it should be opt in - optimise on the extra security/validation first).

File: src/filters/capture.rs
Comment (XAMPPRocky) [reply src/filters/capture.rs:81]: This is not a change in behaviour. See here https://github.com/googleforgames/quilkin/blob/a439516e16199857bda8a2d2a8c3c1bfcf47f37a/src/filters/capture_bytes.rs#L64-L73 the packets are dropped when suffix or prefix don't have enough bytes.

Now whether this behaviour is desirable is a different story. I agree that this filter should probably only be able capturing, and make it configurable as to whether it drops on failing captures. I can add that change to this PR if you agree.

Comment (markmandel) [reply src/filters/capture.rs:81]: ooooh yeah, I see your point. 👍🏻 Oh that's load of fun. Prefix and Suffix are checking for a match just by the nature of checking if there is enough data to actually grab. Excellent point, I didn't think of it that way.

> Now whether this behaviour is desirable is a different story. I agree that this filter should probably only be capturing, and make it configurable as to whether it drops on failing captures. I can add that change to this PR if you agree.

I would absolutely be happy for this addition. I'll leave it up to you if you want to do it in this PR, or we can create an issue and track it as a separate unit of work.

File: docs/src/filters/capture.md
Comment (XAMPPRocky) [reply docs/src/filters/capture.md:24]: > Except Regex can return a Value::List or a single value?

Yes, currently but @iffyio mentioned the alternative being always have it be a list even when it's a single value. 

> (seems to me like it should be opt in - optimise on the extra security/validation first)

This is kinda why I think I'd lean more towards returning a single value over a list with a single value, so that it can be compatible with token router, since most of the you probably only want a single capture group anyway.

I'll update token router's documentation to say that it requires a single value.

File: src/filters/capture.rs
Comment (XAMPPRocky) [reply src/filters/capture.rs:64]: This is now documented in the config struct, once #478 is merged it will be documented in `capture.md`.

**Review by markmandel — APPROVED** [review]

**Pull request merged**
